Add ability to override item's point-range using meta

LandonAConway opened the pull request:
This PR adds the ability to override item point-range via ItemStack meta and fixes #10164 :

`meta:set_string("range", "8")`

## To do

This PR is Ready for Review.

## How to test

In devtest, use the Item Meta Editor to edit a tool, and add the "range" key then set it to a number.

## Notes

For some reason, the files contributed cannot be compared without looking at all of the code. I'm not really sure why this is happening but I'm trying to fix it. If anyone knows of a way to fix it, please let me know.

EDIT: The line endings have been fixed.

appgurueu commented on src/inventory.cpp:
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
+        try {
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'.
+			char *c = const_cast<char *>(sRange.c_str());
```
I don't see why `c` is not a `const char *`, and I don't see why the `const_cast` would be necessary:

* `c_str` has the signature [`const char* c_str() const;`](https://cplusplus.com/reference/string/string/c_str/)
* `strtof` has the signature [`float strtof( const char* str, char** str_end );`](https://en.cppreference.com/w/cpp/string/byte/strtof)

appgurueu commented on src/inventory.cpp:
```diff
@@ -271,6 +271,29 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'
+			f32 _range = std::strtof(sRange.c_str(), NULL);
+            if (_range != 0 || sRange == "0")
```
This looks like a weird way to do validation:

* You will reject something like `0e42` or `0.0`, despite it being a valid float literal.
* It is not clear that the `_range != 0` is actually checking for an error.
* All of this is inside a `try`, implying that this was doing all the error handling (which it doesn't)

Also, the overall logic confuses me: You return range -1 for an "invalid" range, but you return the itemdef range for an empty range. Shouldn't you just default to the itemdef range in both cases?

Side note: Please format this to comply with the [code style guidelines](https://dev.minetest.net/Code_style_guidelines).

LandonAConway replied:
> This looks like a weird way to do validation:
> 
> * You will reject something like `0e42` or `0.0`, despite it being a valid float literal.
> * It is not clear that the `_range != 0` is actually checking for an error.
> * All of this is inside a `try`, implying that this was doing all the error handling (which it doesn't)
> 
> Also, the overall logic confuses me: You return range -1 for an "invalid" range, but you return the itemdef range for an empty range. Shouldn't you just default to the itemdef range in both cases?
> 
> Side note: Please format this entire PR to comply with the [code style guidelines](https://dev.minetest.net/Code_style_guidelines).

I will remove the try-catch block if I can do it the way I'm about to explain. I want to create a function that returns false if the string is not a valid float before converting it. This was my original plan. The function would output like this.

`"Jdh76" -> false`
`"76.9" -> true`
`"836hdg" -> false`
`" " -> false`

Creating this kind of check function would be really simple. The problem is that I would have to place it in the inventory.cpp file which doean't seem like a very organized aproach but rather messy, so then where would I put it?

appgurueu commented on src/inventory.h:
```diff
@@ -51,6 +51,9 @@ struct ItemStack
 	std::string getDescription(const IItemDefManager *itemdef) const;
 	std::string getShortDescription(const IItemDefManager *itemdef) const;
 
+    // Returns the range
```
This comment is pretty redundant.

LandonAConway commented:
I changed the code in `inventory.cpp` so `ItemStack::getRange()` will now have a helper function to check if the string is a valid number so it can be converted to a float and while doing so, it will also convert the string to a float.

I also changed the way `getToolRange()` works (inside `tool.h` and `tool.cpp`) and updated the code which uses that function to comply with the new arguments. I felt that this was neccecary because `getToolRange()` determines which range should be used between the player's hand and the selected item. I.e. if the player has nothing in their hand, then another range will be used. This range from the actual player hand should also use the modified `range` in the ItemStack's metadata. This makes sense and it would not make sense to do it another way. Since `getToolRange()` originally required two arguments of `ItemDefenition`, it could only get the range from those definitions. Instead it now requires two `f32` types representing the player's hand and selected item's ranges which should be retrieved from `ItemStack::getRange()`.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
+        try {
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'.
+			char *c = const_cast<char *>(sRange.c_str());
+			f32 _range = std::strtof(c, NULL);
```
```suggestion
			f32 _range = std::strtof(sRange.c_str(), nullptr);
```
`const_cast` is a special feature and should be treated with caution and respect.

JosiahWI replied:
I would recommend adding the `<cstdlib>` header for `std::strtof`.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
+        range = getDefinition(itemdef).range;
+    if (!sRange.empty())
+    {
+        try {
```
`std::strtof` doesn't throw on parse failure; you don't need this. `std::stof` does throw, however.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
+    f32 range = -1;
+    if (sRange.empty())
+        range = getDefinition(itemdef).range;
+    if (!sRange.empty())
```
```suggestion
    else {
```

JosiahWI commented on src/inventory.cpp:
```diff
@@ -271,6 +271,30 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
+    if (!sRange.empty())
+    {
+        try {
+            //This is the best way (until otherwise said) to convert a string into a float.
+            //std::stof gives a compilation error along the lines of: namespace 'std' does not contain 'mystof'.
```
Did you try including the `<string>` header where `std::stof` is declared?

LandonAConway commented:
Is there a way to remove certian commits from this pr? I updated the line endings settings in .gitattributes but I do not want to push that to official Minetest, and still, my files have the wrong line endings.

JosiahWI commented:
`git rebase` is the tool you are looking for. I recommend using the interactive version (`git rebase -i`), especially if you haven't used the rebase command before.

fluxionary commented:
i haven't played w/ this pull, but i see a potential confound. how does this interact w/ the capabilities of the player's hand, the `""` item?

JosiahWI commented on src/client/game.cpp:
```diff
@@ -3245,7 +3245,7 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud)
 	const ItemStack &tool_item = player->getWieldedItem(&selected_item, &hand_item);
 
 	const ItemDefinition &selected_def = selected_item.getDefinition(itemdef_manager);
```
```suggestion
```
Looks to me like you don't need this temporary anymore.

LandonAConway replied:
I actually did remove it. If you go down further into this function, the `selected_def` reference is actually used. Should it be moved down to where it is used?

JosiahWI replied:
Yes.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -29,6 +29,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/string.h"
 
+//used in ItemStack::getRange()
+static float string_to_float(const std::string& str, bool& is_valid)
+{
+    if (str.empty())
+    {
+        is_valid = false;
+        return 0;
+    }
+
+    char* ptr;
+    float f = strtof(str.c_str(), &ptr);
+    is_valid = (*ptr) == '\0';
+    return f;
+}
```
It'd be a lot cleaner to get `std::stof` working.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -271,6 +286,17 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
 
+f32 ItemStack::getRange(const IItemDefManager *itemdef) const
+{
+    std::string str_meta_range = metadata.getString("range");
```
This should use a reference, probably a const reference.

LandonAConway replied:
As I'm new to C++, I'm not exactly sure what this means. What value are you referring to that should be a const? It seems that they are all const. The function is a const and so is the IItemDefManager.

JosiahWI replied:
I'm sorry, I was unclear. I'm referring to str_meta_range; copying strings is expensive.
```suggestion
    const std::string &str_meta_range = metadata.getString("range");
```

JosiahWI commented on src/inventory.cpp:
```diff
@@ -271,6 +286,17 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
+    f32 range = getDefinition(itemdef).range;
+    bool is_valid;
+    float meta_range = string_to_float(str_meta_range, is_valid);
+    if (is_valid && meta_range > -1)
```
Why is there no warning about the invalid range value?

LandonAConway replied:
Are you asking to have a comment added for a warning?

JosiahWI replied:
I'm asking why nothing is printed to the log. Invalid meta seems like something that should generate at least a warning.

LandonAConway commented:
> `git rebase` is the tool you are looking for. I recommend using the interactive version (`git rebase -i`), especially if you haven't used the rebase command before.

I see exactly what you are talking about and it has been fixed already to handle this. (should be).
In `tool.cpp` there is a function that is called `getToolRange()`. It has been changed to take two floats (one for the selected item, and one for the hand) and returns the one to be used. This function once took two item definitions (one for the selected item, and hand) then got the range from the definitions but it has been changed to evaluate two floats. Both floats are retrieved from `ItemStack::getRange()` which gets the item definition's range, and if the metadata has a value at the key of `"range"` that is convertable to a float, it uses this instead.

Desour commented:
The diffs are broken.

Also docs are missing.

LandonAConway commented:
> The diffs are broken.
> 
> Also the documentation is missing.

I will get back to this PR hopefully in a couple days. I'm busy at work. I WILL be working on it so if there is inactivity for a couple days, just beware. This PR is really important to me.

JosiahWI commented:
serverpackethandler.cpp has been severely afflicted by these changes.

LandonAConway commented:
> serverpackethandler.cpp has been severely afflicted by these changes.

Literally only one line has changed. I have updated ALL the file's line endings so hopefully that will work.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -29,6 +29,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+    if (str.empty())
+    {
+        is_valid = false;
+        return 0.0;
```
`0.0f`, `0.0` is a `double` literal.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -271,6 +286,19 @@ std::string ItemStack::getShortDescription(const IItemDefManager *itemdef) const
 
+f32 ItemStack::getRange(const IItemDefManager *itemdef) const
+{
+    std::string str_meta_range = metadata.getString("range");
```
There is still an unnecessary string copy here.

JosiahWI commented on src/inventory.cpp:
```diff
@@ -29,6 +29,21 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #include "util/string.h"
 
+//used in ItemStack::getRange()
+static float string_to_float(const std::string& str, bool& is_valid)
```
We can use `std::optional` now that we have C++17 support, instead of having a boolean out parameter.

LandonAConway commented:
I'm aware of the changes needed. I willbe working on them this week (hopefully).
